### PR TITLE
MAINT: Move `RankWarning` to exceptions module

### PR DESCRIPTION
**Pull request by mtsokol**
Connected to #23999 

Hi @rgommers @ngoldbaum,

This PR moves `RankWarning` exception from `np.polynomial` to `np.exceptions` and removes it from main namespace. Other closely related libraries aren't affected.

### CODE REVIEW

File: numpy/exceptions.py [hunk 1]
```diff
@@ -79,6 +81,16 @@ class VisibleDeprecationWarning(UserWarning):
     the usage is most likely a user bug.
 
     """
+    pass
+
+
+class RankWarning(RuntimeWarning):
+    """Martrix rank warning.
```
Comment (rgommers) [inline numpy/exceptions.py:88]: typo in Matrix

**Review by rgommers — APPROVED** [review]
Interesting, so we had two separate ones:
```python
>>> np.polynomial.polyutils.RankWarning is np.RankWarning
False
```

Makes sense to me to reduce that to one, and put it in `numpy.exceptions`. 

LGTM modulo the one typo.

File: numpy/exceptions.py (hunk 1 above)
Comment (mtsokol) [reply numpy/exceptions.py]: Ah, sure - fixed!

**Comment by rgommers** [comment]
No more comments, in it goes. Thanks Mateusz.

**Pull request merged**